libs/cangen/src/can_gen_decode.rs

```diff
@@ -0,0 +1,213 @@
+            }
+            // Otherwise, send it (default)
+            _ => {
+                let unit = self.unit.clone();
+
+                // If topic_append, we need to set up the suffix
+                match self.topic_append {
+                    Some(true) => {
+                        let mut topic_suffix_point = self.points.remove(0);
+                        let topic_suffix_read = topic_suffix_point.gen_decoder_fn();
+                        let topic = format!("{}/", self.name);
+                        let point_decoders = self
+                            .points
+                            .iter_mut()
+                            .map(|point| {
+                                let decoder = point.gen_decoder_fn();
+                                quote! { decoded_points.push(#decoder); }
+                            })
+                            .collect::<Vec<ProcMacro2TokenStream>>()
+                            .into_iter()
+                            .fold(ProcMacro2TokenStream::new(), |mut acc, ts| {
+                                acc.extend(ts);
+                                acc
+                            });
+                        quote! {
+                            topic_suffixes.push(#topic_suffix_read);
+                            #point_decoders
+                            result.push(
+                                DecodeData::new(decoded_points.clone(),
+                                    #topic + String::from(topic_suffixes.pop()),
+                                    #unit)
+                            );
+                            decoded_points.clear();
+                        }
+                    }
+                    _ => {
+                        let point_decoders = self
+                            .points
+                            .iter_mut()
+                            .map(|point| {
+                                let decoder = point.gen_decoder_fn();
+                                quote! { decoded_points.push(#decoder); }
+                            })
+                            .collect::<Vec<ProcMacro2TokenStream>>()
+                            .into_iter()
+                            .fold(ProcMacro2TokenStream::new(), |mut acc, ts| {
+                                acc.extend(ts);
+                                acc
+                            });
+                        let topic = self.name.clone();
+                        quote! {
+                            #point_decoders
+                            result.push(
+                                DecodeData::new(decoded_points.clone(),
+                                    #topic,
+                                    #unit)
+                            );
+                            decoded_points.clear();
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn gen_decoder_skip(&mut self) -> ProcMacro2TokenStream {
+        quote! { }
```

Comment:
just curious why this is empty?

Reply:
Yeah this was kind of an obtuse design decision, I'll refactor the traits in `can_gen_decode` and `can_gen_encode` 

Original idea was to encapsulate the functionality of generating functions into a single trait, but it seems better to handle this either in the appropriate macro helper (for mappings and key list) or in `gen_decoder_fn` (in the case of `gen_decoder_skip`)